### ISSUE TITLE
tcptracer invalid mem access on older Linux

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bcc#1482]
On Linux 4.9-rc2:

```
# ./tcptracer.py
bpf: Permission denied
0: (79) r7 = *(u64 *)(r1 +80)
1: (85) call 14
2: (7b) *(u64 *)(r10 -8) = r0
3: (18) r1 = 0x4e0d6540
5: (bf) r2 = r10
6: (07) r2 += -8
7: (85) call 1
8: (bf) r6 = r0
9: (15) if r6 == 0x0 goto pc+55
 R0=map_value_or_null(ks=8,vs=8),min_value=0,max_value=0 R6=map_value(ks=8,vs=8),min_value=0,max_value=0 R7=inv R10=fp
10: (18) r1 = 0x4e0d6540
12: (bf) r2 = r10
13: (07) r2 += -8
14: (85) call 3
15: (67) r7 <<= 32
16: (77) r7 >>= 32
17: (55) if r7 != 0x0 goto pc+47
 R0=inv,min_value=0,max_value=0 R6=map_value(ks=8,vs=8),min_value=0,max_value=0 R7=inv32,min_value=0,max_value=0 R10=fp
18: (79) r1 = *(u64 *)(r6 +0)
19: (b7) r2 = 0
20: (7b) *(u64 *)(r10 -16) = r2
21: (7b) *(u64 *)(r10 -24) = r2
22: (63) *(u32 *)(r10 -48) = r2
23: (69) r6 = *(u16 *)(r1 +12)
R1 invalid mem access 'inv'

HINT: The invalid mem access 'inv' error can happen if you try to dereference memory without first using bpf_probe_read() to copy it to the BPF stack. Sometimes the bpf_probe_read is automatic by the bcc rewriter, other times you'll need to be explicit.

Traceback (most recent call last):
  File "./tcptracer.py.1", line 628, in <module>
    b.attach_kretprobe(event="tcp_v4_connect", fn_name="trace_connect_v4_return")
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 552, in attach_kretprobe
    fn = self.load_func(fn_name, BPF.KPROBE)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 346, in load_func
    (func_name, errstr))
Exception: Failed to load BPF program trace_connect_v4_return: Permission denied
```

But it works on Linux 4.15-rc2.

If the fix was trivial, I'd fix tcptracer to work on Linux 4.9 as well, since it's LTS. But if it was anything other than trivial, I'd be inclined to just note that it needs a newer Linux kernel version in both the tool source and the man page. See the "REQUIREMENTS" section in profile.py and profile.8 as an example (profile.py needs 4.9+).

**Comment by iaguis** [comment]
Current master worked for me on 4.9.67. However, I pushed a fix for older kernels in #1483. It worked for me in 4.4.105 so I guess it should also work for you.

**Comment by yonghong-song** [comment]
Agreed with @iaguis. Maybe you do not have latest bcc repo on 4.9?
With the latest master,
```
0: (bf) r6 = r1
1: (79) r7 = *(u64 *)(r6 +80)
2: (85) call bpf_get_current_pid_tgid#14
3: (bf) r8 = r0
4: (15) if r7 == 0x0 goto pc+162
 R0=inv(id=0) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv(id=0) R10=fp0
5: (7b) *(u64 *)(r10 -120) = r6
6: (b7) r6 = 0
7: (6b) *(u16 *)(r10 -2) = r6
8: (6b) *(u16 *)(r10 -4) = r6
9: (63) *(u32 *)(r10 -8) = r6
10: (bf) r3 = r7
11: (07) r3 += 12
12: (bf) r1 = r10
13: (07) r1 += -4
14: (b7) r2 = 2
15: (85) call bpf_probe_read#4
16: (bf) r3 = r7
17: (07) r3 += 14
18: (bf) r1 = r10
19: (07) r1 += -2
20: (b7) r2 = 2
21: (85) call bpf_probe_read#4
...
```
bpf_probe_read is used properly to avoid the error you saw in the above.
